src/kernels/gpu_batched_transpose_kernel/batched_transpose.cpp

```diff
@@ -0,0 +1,2066 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2020-2021 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#include <hip/hip_runtime.h>
+#include <hip/hip_fp16.h>
+
+#ifndef BATCHED_TRANSPOSE_OCCUPANCY
+#define BATCHED_TRANSPOSE_OCCUPANCY 4
+#endif
+
+inline __device__ uint32_t magic_div_u32(const uint32_t& numer,
+                                         const uint32_t& magic,
+                                         const uint32_t& shift)
+{
+    uint32_t tmp = __umulhi(numer, magic);
+    return (tmp + numer) >> shift;
+}
+
+inline __device__ void v_pack_b32_f16_00(float& c, const float& a, const float& b)
+{
+#if 0
+    asm volatile("v_pack_b32_f16 %0, %1, %2\n"
+                 : "=v"(c)
+                 : "v"(a), "v"(b));
+#else
+    const uint32_t x  = *reinterpret_cast<const uint32_t*>(&a);
```

Comment:
@carlushuang @atamazov  HIP tidy is complaining about lots of casts in this file. If indeed they are necessary, let's silence the warning on `[invalidPointerCast]` for this file?

ref: [HIP tidy log](http://micimaster.amd.com/blue/rest/organizations/jenkins/pipelines/MLLibs/pipelines/MIOpen/branches/nchw_xdlops_support_by_transpose/runs/8/nodes/13/steps/63/log/?start=0#:~:text=kernels%20batch%20%2366-,%5B2021-11-03T10%3A44%3A22.913Z%5D%20/home/jenkins/workspace/nchw_xdlops_support_by_transpose/src,due%20to%20different%20binary%20data%20representations%20on%20different%20platforms.%20%5BinvalidPointerCast%5D,-%5B2021-11-03T10)

Reply:
@junliume try workaround it by adding `// cppcheck-suppress invalidPointerCast`